x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTask.java

```diff
@@ -5,90 +5,393 @@
+        }
+
+        LOGGER.trace("{} coordinate reads, lastRequestedSeqno={}, globalCheckpoint={}",
+            params.getFollowShardId(), lastRequestedSeqno, globalCheckpoint);
+        final int maxBatchOperationCount = params.getMaxBatchOperationCount();
+        while (hasReadBudget() && lastRequestedSeqno < globalCheckpoint) {
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            long maxRequiredSeqno = Math.min(globalCheckpoint, from + maxBatchOperationCount);
+            LOGGER.trace("{}[{}] read [{}/{}]", params.getFollowShardId(), numConcurrentReads, maxRequiredSeqno, maxBatchOperationCount);
+            sendShardChangesRequest(from, maxBatchOperationCount, maxRequiredSeqno);
+            lastRequestedSeqno = maxRequiredSeqno;
+        }
+
+        if (numConcurrentReads == 0 && hasReadBudget()) {
+            assert lastRequestedSeqno == globalCheckpoint;
+            // We sneak peek if there is any thing new in the leader.
+            // If there is we will happily accept
+            numConcurrentReads++;
+            long from = lastRequestedSeqno + 1;
+            LOGGER.trace("{}[{}] peek read [{}]", params.getFollowShardId(), numConcurrentReads, from);
+            sendShardChangesRequest(from, maxBatchOperationCount, lastRequestedSeqno);
+        }
+    }
+
+    private boolean hasReadBudget() {
+        assert Thread.holdsLock(this);
+        if (numConcurrentReads >= params.getMaxConcurrentReadBatches()) {
+            LOGGER.trace("{} no new reads, maximum number of concurrent reads have been reached [{}]",
+                params.getFollowShardId(), numConcurrentReads);
+            return false;
+        }
+        if (buffer.size() > params.getMaxWriteBufferSize()) {
+            LOGGER.trace("{} no new reads, buffer limit has been reached [{}]", params.getFollowShardId(), buffer.size());
+            return false;
+        }
+        return true;
+    }
+
+    private synchronized void coordinateWrites() {
+        while (hasWriteBudget() && buffer.isEmpty() == false) {
+            long sumEstimatedSize = 0L;
+            int length = Math.min(params.getMaxBatchOperationCount(), buffer.size());
+            List<Translog.Operation> ops = new ArrayList<>(length);
+            for (int i = 0; i < length; i++) {
+                Translog.Operation op = buffer.remove();
+                ops.add(op);
+                sumEstimatedSize += op.estimateSize();
+                if (sumEstimatedSize > params.getMaxBatchSizeInBytes()) {
+                    break;
+                }
+            }
+            numConcurrentWrites++;
+            LOGGER.trace("{}[{}] write [{}/{}] [{}]", params.getFollowShardId(), numConcurrentWrites, ops.get(0).seqNo(),
+                ops.get(ops.size() - 1).seqNo(), ops.size());
+            sendBulkShardOperationsRequest(ops);
+
+            // In case that buffer is higher than max write buffer size then reads may all have been stopped,
```

Comment:
it feels a bit weird to have this here, in the coordinateWrite loop. Since it's always safe to call coordinateReads, how about just calling it once in handleWriteResponse and be done with it?